settings_urls.txt
# --Syntax to use for missing values (as a Python string):
missing_value = "n/d"


# ############################################################################
# ############################################################################
# ### DIRECTORIES AND FILE NAMES:
# ##################################
#
# 0. Path where the exported JSON files are (as a Python string):
jsons_source_path = "/home/agds/Documents/RET/Source/Jan_1_2025_to_Jan_31_2025/"
#
# 1. Path where the exported Excel files are (as a Python string):
excel_channels_path = "/home/agds/Documents/RET/Source/Jan_1_2025_to_Jan_31_2025/_JSONs_converted"
#
# 2. Name of the Slack channel dedicated to the weekly reports (as a Python string):
reports_channel_name = "think-biver-weekly-checkins"
#
#
urls_to_show = ["docs.google.com", "figma.com"] 
#
# 3. Name of file to be saved with the compilation of messages with URLs (as a Python string):
compilation_urls_file_name = "urls_2.xlsx"
#
# 4. Path where the file compilation_reports_file_path will be saved (as a Python string):
compilation_urls_path = "/home/agds/Desktop"



# ############################################################################
# ############################################################################
# ### EXCEL FORMATTING:
# #####################
#
# 0. Set the order of the columns in the Excel table (as a Python list of strings):
columns_order = ["display_name", "URL(s)", "msg_date", "channel"]
#
# 1. Change the width of the Excel cells (as Python intege numbers and Python dictionary):
w_date = 19
w_name = 19
w_text = 30
w_bool = 7
w_count = 8
column_widths = {
    "channel": w_text, "msg_date": w_date,
    "display_name": w_name, "text": w_text, "URL(s)": w_text
}
# -----------------------------------------------------------------------------
# 2. Set the alignment of cells containing long text (as a Python list of strings):
text_type_cols = ["B"]
#
# -----------------------------------------------------------------------------
# 3. Format the height, color, font alignment and size of the first column (as a Python dictionary):
header_row = {
    "height": 43,
    "alignment_vert": "top",
    "alignment_horiz": "left",
    "font_size": 9,
    "font_bold": True,
    "cell_color": [
        # ("color_code", [list the columns' letter])
        ["e7c9fb", ["A", "B", "C", "D"]]
        ]
}


settings_stats.txt
# --Syntax to use for missing values (as a Python string):
missing_value = "n/d"


# ############################################################################
# ############################################################################
# ### DIRECTORIES AND FILE NAMES:
# ##################################
#
# 0. Path where the exported JSON files are (as a Python string):
jsons_source_path = "/home/agds/Documents/RET/Source/Jan_1_2025_to_Jan_31_2025/"
#jsons_source_path = r"C:\Users\angel\OneDrive\Documents\RebeccaEverleneTrust\RET_source\Jan_1_2025-Jan_31_2025\"
#
# 1. Path where the exported Excel files are (as a Python string):
excel_channels_path = "/home/agds/Documents/RET/Source/Jan_1_2025_to_Jan_31_2025/_JSONs_converted"
#excel_channels_path = r"C:\Users\angel\OneDrive\Documents\RebeccaEverleneTrust\RET_source\Jan_1_2025-Jan_31_2025\_JSONs_converted"
#
# 2. Name of the Slack channel dedicated to the weekly reports (as a Python string):
reports_channel_name = "think-biver-weekly-checkins"
#
# 3. Name of file to be saved with the compilation of check-in weekly reports (as a Python string):
compilation_reports_file_name = "test_reports_v4.xlsx"
#
# 4. Path where the file compilation_reports_file_path will be saved (as a Python string):
compilation_reports_path = r"/home/agds/Desktop"
#compilation_reports_path = r"C:\Users\angel\Desktop"



# ############################################################################
# ############################################################################
# ### EXCEL FORMATTING:
# #####################
#
# 0. Set the order of the columns in the Excel table (as a Python list of strings):
columns_order = ["channel", "export_dates", "parsed_reports_in_channel",
                 "user", "name", "display_name", "deactivated", "is_bot",
                 "msg_id", "msg_date", "type", "text",
                 "reply_count", "reply_users_count", "latest_reply_date",
                 "thread_date", "parent_user_name", "URL(s)",
                 "number_msgs_in_channel", "projects_parsed",
                 "keywords_parsed", "latest_report_date", "project_name",
                 "working_on", "progress_and_roadblocks", "progress",
                 "roadblocks", "plans_for_following_week", "meetings"]
#
# 1. Change the width of the Excel cells (as Python intege numbers and Python dictionary):
w_date = 19
w_name = 19
w_text = 30
w_bool = 7
w_count = 8
column_widths = {
    "channel": w_text, "export_dates": 25, "parsed_reports_in_channel": w_count,
    "msg_id": 12, "msg_date": w_date, "user": 15, "name": w_name,
    "display_name": w_name, "deactivated": w_bool, "is_bot": w_bool,
    "type": 8, "text": w_text, "reply_count": w_count,
    "reply_users_count": w_count, "latest_reply_date": w_date,
    "thread_date": w_date, "parent_user_name": w_name,
    "URL(s)": w_text, "number_msgs_in_channel": w_count,
    "projects_parsed": w_count, "keywords_parsed": w_count,
    "project_name": w_text, "working_on": w_text,
    "progress_and_roadblocks": w_text, "progress": w_text,
    "roadblocks": w_text, "plans_for_following_week": w_text,
    "meetings": w_text, "latest_report_date": w_date
}
# -----------------------------------------------------------------------------
# 2. Set the alignment of cells containing long text (as a Python list of strings):
text_type_cols = ["L", "W", "X", "Y", "Z", "AA", "AB", "AC"]
#
# -----------------------------------------------------------------------------
# 3. Format the height, color, font alignment and size of the first column (as a Python dictionary):
header_row = {
    "height": 43,
    "alignment_vert": "top",
    "alignment_horiz": "left",
    "font_size": 9,
    "font_bold": True,
    "cell_color": [
        # ("color_code", [list the columns" letter])
        ["e7c9fb", ["A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R"]],
        ["CDB5B7", ["C", "S", "T", "U", "V", "W", "X", "Y", "Z", "AA", "AB", "AC"]]
        ]
}
# -----------------------------------------------------------------------------
# 4. Distinguish special columns by changing the color of their font (as a Python list of lists):
font_color_in_column = [
    # ["color_code", "column_letter"]
    ["0707C5", "I"],  # "text" to blue
    ["c10105", "O"],  # "parent_user_name" to red
    ["c10105", "E"]   # "display_name" to red
]
# -----------------------------------------------------------------------------
# 5. Change the color and/or font on chosen cells based on a certain condition
# you wish to highlight (as a Python dictionary):
highlights = [
    # 5.0. Highlight format:
    # {"activate": True/False,
    # "trigger": ["column_letter", "condition", value],
    # "columns": [list of columns to highlight],
    # "cell_color": "color_code",
    # "font_size": #,
    # "font_bold": True/False,
    # "font_horiz_alignment": "alignment_type"
    # },
    # 5.1. Highlight cases where the user is a bot.
    {"activate": True,
    "trigger": ["H", "==", True],
    "columns": ["D", "E", "F", "G", "H", "S", "T", "U", "V", "W", "X", "Y", "Z", "AA", "AB", "AC"],
    "cell_color": "FBBF8F",
    "font_size": 11,
    "font_bold": False,
    "font_horiz_alignment": "left"
    },
    # 5.2. Highlight cases where the message is part of a thread.
    {"activate": True,
    "trigger": ["K", "==", "thread"],
    "columns": ["K", "L", "S", "T", "U", "V", "W", "X", "Y", "Z", "AA", "AB", "AC"],
    "cell_color": "FBFB99",
    "font_size": 11,
    "font_bold": False,
    "font_horiz_alignment": "left"
    },
    # 5.3. Highlight the cases where there is reply_counts.
    {"activate": True,
    "trigger": ["M", "!=", "n/d"],
    "columns": ["M", "N"],
    "cell_color": "No Fill",
    "font_size": 12,
    "font_bold": True,
    "font_horiz_alignment": "right"
    }
]
# -----------------------------------------------------------------------------
# 6. Draw vertical lines at the right of the given columns (as a Python dictionary):
draw_vert_line = {
    "columns": ["C", "H", "R"],
    "thickness": "medium"
    }


CHANGELOG.md
# Change Log

**03/16/2025**
* Updated README and CHANGELOG.
* Unified the structure of settings file for header_row and font_color_in_column.
* Added name of reports channel to settings_stats file.
* Added sheet_name when reading the channels Excel files.
* Added pruning of channel messages. Filtered-out msgs are kept in a different Excel tab. 
* Renamed variables in `messages.py` to simplify visual appearance of the code.

**03/08/2025**
* Added tab with unparsed weekly reports from (exclusively) the weekly-reports Slack-channel. 
* Added filters in Excel table with compiled weekly reports, allowing filtering and/or sorting of all the columns.
* Added function to draw a vertical line in an Excel table, given the specs of `draw_vert_line` in the `settings_stats.py`.
* Added check of the channels' names when reading the Excel tables with the exported messages.
* Added variables to `settings_messages.py`: `channels_excel_name` and `users_excel_name` to be used as names of the corresponding files.

**03/04/2025**
* Added worksheet as argument on Excel functions and created tabs of by_channel and by_user in the compiled Excel file.

**03/02/2025**
* Improved the coding structure of the `stats.py` file and changed its location from `/src/stats.py` to `/VScode/stats.py`.
* Renamed the settings file needed to generate the Excel files for each Slack channel from `settings.py` to `settings_messages.py`. (To be more explicit and avoid confusion with the `settings_stats.py` file)
* The files with the user's inputs are now arguments that need to be specified when running the Python scripts. This required a few structural changes to the customized modules and the Jupyter notebook variant.
The changes gives more flexibility when using the various modules and makes the code easily generalizable if it is needed to generate a new Excel file with different information. 
The commands to be run in the terminal are now (the paths can be relative or absolute):
  - python main.py --inputs_file_path="..\inputs.py" --seettings_file_path="..\settings_messages.py"
  - python stats.py  --seettings_file_path="..\settings_stats.py"
* Updated the documentation.

**02/26/2025:**
* Replaced empty spaces with a dash when naming the Excel files in `main.py`. This is only a problem for channels starting with FC_.
* Added first draft of Excel file with stats on weekly reports. The main script is in `src/stats.py` and the excel settings are in `settings_stats.py`. 
Needed to generalize some aspects of the `excel.py` implementation to make it flexible and usable for both `main.py` and `stats.py`.
* More exceptions were added to the reports parser.

**02/21/2025:**
Changed the structure of the git repo:
* Input files `inputs.py` and `settings.py` are in the main repo directory.
* There are now three folders for the variants: VScode, JupyterNotebook, and GUI.
* `./VScode` has the Python script `main.py` and documentation on how to use it in VScode.
* `./JupyterNotebook` has the notebook `main.ipynb`, which contains its documentation.
* `./GUI` has the executable file slack2excel.exe, the python script `GUI_tkinter.py` with the main code of the graphic interface, and the subdirectory `tkGUI`, 
which is the output from generating the one-file executable app with pyinstaller.
* The requirements.txt and the bash scripts were collected in the directory `./dependencies`.
* Screenshots used to build documentation are stored in the directory `./images`.
* Updated the README file.
* Added a CHANGELOG file with some of the previous updates.
* The new file structure required changing how specific modules were imported in `main.py` and `GUI_tkinter.py`.

**02/14/2025:**
Extracted check-in variables and Excel variables to the file `settings.py`.
* The check-in variables in `settings.py` are now: all_keywords, index_keyword, keywords_dictionary, sample_text_list. Extracted from `checkins.py` and `excel.py`.
* The excel variables in `settings.py` are now: column_width, text_type_cols, height_1strow, alignment_vert_1strow, alignment_horiz_1strow, font_size_1strow, font_bold_1strow, cell_color_1strow, font_color_in_column, highlights.
* The Excel variables were all extracted from `excel.py`. It required generalizing the its internal functions, as well as adopting specific Python structures for the variables in `settings.py`.
* The `settings.py` file is intended for use by developers or technical users.

**02/12/2025:**
* Fixed a bug preventing the output directory from being refreshed every time.
* Recover some Excel formatting that was lost in the jupyter-modules translation.

**02/11/2025:**
* The parsing of check-in messages is now applied to all the channels. There is no input variable asking for the name of the check-ins channel anymore (there was a left-over hardcoded checkin_channel_name variable in `messages.py`).
* Added 'missing_value' instead of empty string to empty rows in the URL column. Now all the cells with no information have 'n/d'.

**02/08/2025:**
* GUI now contains visual cues and a help text box to guide the user using the app.
* Changed the checkbuttons for the channels_flag and users_flag to radiobuttons. There is no default value anymore, user have to enter either 'yes' or 'no'.

**2025/01/31:**
* Minor style changes.
* Added channels_json_name, users_json_name, checkin_channel_name, and dest_name_ext to `settings.py`.


settings.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# --Syntax to use for missing values:
missing_value = 'n/d'

# --Specify time-zone region to express the date:
timezone = 'US/Central'


# ############################################################################
# ############################################################################
# ### DIRECTORIES AND FILE NAMES:
# ##################################
#
# 1. Name of file with the channels information located in the source directory:
channels_json_name = 'channels.json'
#
# -----------------------------------------------------------------------------
# 2. Name of file with the users information located in the source directory:
users_json_name = 'users.json'
#
# -----------------------------------------------------------------------------
# 3. Name extension to the destination folder:
dest_name_ext = '_JSONs_converted'


# ############################################################################
# ############################################################################
# ### KEYWORDS OF CHECK-IN MESSAGES:
# ##################################
#
# 1. List of keywords to look for in each check-in message:
# (It is assumend that the text follows the convention "keyword : ")
all_keywords = ['project_name', 'working_on', 'progress_and_roadblocks',
                'progress', 'roadblocks', 'plans_for_following_week',
                'meetings']
#
# -----------------------------------------------------------------------------
# 2. Which keyword do you want to use as trigger for distinguishing different
# projects?
# NOTE: currently chosing project_name since its use is more consistent than
# 'weekly report:'
index_keyword = 'project_name'
#
# -----------------------------------------------------------------------------
# 3. To maximize the parsing of messages, enter words/phrases that users are
# likely to use as keywords in their check-in messages.
# Enter your inputs in the form of a Python dictionary as shown bellow:
keywords_dictionary = {
    # 'keyword': ['keyword/phrase 1', ..., 'keyword/phrase n'],
    'header':
        ['weekly report', 'report', "week's report"],
    'project_name':
        ['project name', 'Current Project Name', 'project'],
    'working_on':
        ['working on', 'working', 'worked on', "What I'm working on",
         'What you are working', 'what you are working on',
         'current task', 'current_tasks', 'task', 'current focus'],
    'progress_and_roadblocks':
        ['progress and roadblocks', 'progress and roadblock',
        'progress &amp; roadblocks', 'Progress/Roadblocks',
        'progress and challenges'],
    'progress':
        ['progress'],
    'roadblocks':
        ['roadblocks', 'roadblock'],
    'plans_for_following_week':
        ['plans for the following week', 'plans for next week',
         'Plans for the next week', 'following week', 'next week',
         'plans for the upcoming week', 'Plans for the weeks',
         'upcoming plans', 'Plans For Following Week',
         'Plans for upcoming week', 'plan for next week'],
    'meetings':
        ['meetings', 'meet', 'met', "meetings you've attended",
        'upcoming meetings', 'meeting', 'Meeting attended', 'Meetings attended']
}
# -----------------------------------------------------------------------------
# 4. Some of the checkin messages are sent to the Slack channel as a sample on
# how to properly format the expected message. In the list sample_text_list
# bellow, enter a fragment of the text of such "sample" messages. This list
# will be use to explicitely identifying the sample messages in the final
# Excel sheet:
sample_text_list = [
    "We’re working on a volunteer tracking project for HR, centralizing all information in Salesforce",
    "<!channel> reposting <@U07FCQXU7Q9>'s message. Please adhere to it. THANK YOU.",
    "please follow this structure when posting updates",
    "Reminder: Hey everyone! If you haven't already, don't forget to submit your weekly check-in no later than Saturday 11:59pm local time",
    "should include a summary of your accomplishments, ETC. (follow the outline below)"
]

# ############################################################################
# ############################################################################
# ### EXCEL FORMATTING:
# #####################
#
# 1. Change the width of the Excel cells:
w_date = 19
w_name = 19
w_text = 30
w_bool = 7
w_count = 8
column_widths = {
    'msg_id': 12, 'msg_date': w_date, 'user': 15, 'name': w_name,
    'display_name': w_name, 'deactivated': w_bool, 'is_bot': w_bool,
    'type': 8, 'text': w_text, 'reply_count': w_count,
    'reply_users_count': w_count, 'latest_reply_date': w_date,
    'thread_date': w_date, 'parent_user_name': w_name,
    'URL(s)': w_text, 'projects_parsed': w_count, 'keywords_parsed': w_count,
    'project_name': w_text, 'working_on': w_text,
    'progress_and_roadblocks': w_text, 'progress': w_text,
    'roadblocks': w_text, 'plans_for_following_week': w_text,
    'meetings': w_text
}
# -----------------------------------------------------------------------------
# 2. Set the alignment of cells containing long text:
text_type_cols = ['I', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X']
#
# -----------------------------------------------------------------------------
# 3. Format the height, color, font alignment and size of the first column:
height_1strow = 43
alignment_vert_1strow = "top"
alignment_horiz_1strow = "left"
font_size_1strow = 9
font_bold_1strow = True
cell_color_1strow = [
    # ("color_code", [list the columns' letter])
    ("e7c9fb", ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O']),
    ("CDB5B7", ['P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X'])
]
# -----------------------------------------------------------------------------
# 4. Distinguish special columns by changing the color of their font:
font_color_in_column = [
    # ["column_letter', 'color_code']
    ('I', "0707C5"),  # "text" to blue
    ('N', "c10105"),  # "parent_user_name" to red
    ('E', "c10105")   # "display_name" to red
]
# -----------------------------------------------------------------------------
# 5. Change the color and/or font on chosen cells based on a certain condition
# you wish to highlight:
highlights = [
    # 5.0. Highlight format:
    # {"activate": True/False,
    # "trigger": ["column_letter", "condition", value],
    # "columns": [list of columns to highlight],
    # "cell_color": "color_code",
    # "font_size": #,
    # "font_bold": True/False,
    # "font_horiz_alignment": "alignment_type"
    # },
    # 5.1. Highlight cases where the user is a bot.
    {"activate": True,
    "trigger": ["G", "==", True],
    "columns": ["C", "D", "E", "F", "G", "P", "Q", "R", "S", "T", "U", "V", "W", "X"],
    "cell_color": "FBBF8F",
    "font_size": 11,
    "font_bold": False,
    "font_horiz_alignment": "left"
    },
    # 5.2. Highlight cases there the message is part of a thread.
    {"activate": True,
    "trigger": ["H", "==", "thread"],
    "columns": ["H", "I", "P", "Q", "R", "S", "T", "U", "V", "W", "X"],
    "cell_color": "FBFB99",
    "font_size": 11,
    "font_bold": False,
    "font_horiz_alignment": "left"
    },
    # 5.3. Highlight the cases where there is reply_counts.
    {"activate": True,
    "trigger": ["J", "!=", "n/d"],
    "columns": ["J", "K"],
    "cell_color": "No Fill",
    "font_size": 12,
    "font_bold": True,
    "font_horiz_alignment": "right"
    }
]
#


# --Ready to start the analysis?:
continue_analysis = True